lib/puppet/util/execution.rb

```diff
@@ -89,6 +89,47 @@ def self.execpipe(command, failonfail = true)
     output
   end
 
+  # Wait for a process to exit or kill it after a given timeout
+  #
+  # This function polls for the exit status of a child process. If the
+  # supplied timeout is exceeded, then the child is terminated.
+  #
+  # @param child_pid [Integer] The PID of the process to monitor.
+  # @param timeout [Integer] The maximum amount of time to wait, in seconds.
+  #
+  # @return [Array<Integer, Process::Status>] A two element array containing
+  #   the PID of the monitored process along with a Process::Status object
+  #   containing its exit status as returned by Process.waitpid2.
+  def self.wait_with_timeout(child_pid, timeout)
+    # TODO: Use a monotonic clock.
+    deadline = Time.now + timeout
+    # The Process module on Windows only supports :KILL, unless a win32
+    # extension gem is used. Even then, the :TERM supplied is equivalent
+    # to :KILL
+    kill_signal = Puppet.features.microsoft_windows? ? :KILL : :TERM
+
+    loop do
+      # Poll status of child process without blocking. Returns nil
+      # if the process has not exited.
+      status = Process.waitpid2(child_pid, Process::WNOHANG)
+
+      return status unless status.nil?
+
+      # TODO: Use a monotonic clock.
+      if Time.now > deadline
+        Puppet.err(_('Child process exceded timeout. Sending %{signal} signal to PID %{pid}') %
+                   {pid: child_pid,
+                    signal: kill_signal})
+
+        Process.kill(kill_signal, child_pid)
```

Comment:
Note: win32-process implements its own `kill` at https://github.com/djberg96/win32-process/blob/4babb9e410fe68949bc03ee262aa67927637dd23/lib/win32/process.rb#L743

Reply:
What is the expected behavior of `Process.kill` if the process finished since calling `waitpid2`?

Reply:
It returns 0, unless something has reaped the process by a `waitpid2` call that successfully returned a `Process::Status`. The assumption here is that this routine is the only thing calling `waitpid2`.

Reply:
If the process does not exist, the documentation for [kill(3)](https://linux.die.net/man/3/kill) says it should return ESRCH. AFAICT Ruby does not handle this specially, so it will lead to a standard Ruby system call failure exception with errno set to ESRCH. We should handle this case.

Reply:
Yeah, it's a bit tricky since this method expects to reap the process and return a `Process::Status`, so simply handling `ESRCH` doesn't really resolve the issue if some other thread is reaping the child PID. Which is its self an issue since Puppet is mostly single-threaded. I can't think of a case where we'd intentionally call waitpid2 on something twice from two threads. There are cases where we explicitly spawn reaper threads when we don't care about the exit status and just want to ensure something cleans up the child, but I can't imagine we would also use `wait_with_timeout` in addition to a reaper thread.

However, I'm re-factoring this entire behavior to use Timeout.timeout instead of sending kills, so this will likely become a moot point if the refactor is successful.